torch/_dynamo/variables/torch.py

```diff
@@ -1444,6 +1444,86 @@ def handle_check(self, tx: "InstructionTranslator", *args, **kwargs):
+                    ],
+                )
+
+            tx.output.uses_autograd_grad = True
+
+            # Graph break if any input tensor has grad_fn created outside the current graph.
+            # We can't trace through external grad_fn since it's not in our FX graph.
+            # Two checks needed: (1) lazy args not yet in input_source_to_var,
+            # (2) already-realized graph inputs in input_source_to_var.
+
+            def check_lazy_tensor_grad_fn(var):
+                if isinstance(var, LazyVariableTracker) and not var.is_realized():
+                    value = var.peek_value()
+                    if isinstance(value, torch.Tensor) and value.grad_fn is not None:
+                        unimplemented(
+                            gb_type="autograd.grad with external grad_fn (arg)",
+                            context="",
+                            explanation=(
+                                "torch.autograd.grad() cannot be used when the compiled function "
+                                "receives tensors with grad_fn created outside the compiled region. "
+                                "The autograd graph cannot extend beyond the compiled region boundary."
+                            ),
+                            hints=[
+                                "Ensure all tensor inputs to the compiled function are leaf tensors.",
+                                "Or move the autograd.grad() call outside the compiled region.",
+                                *graph_break_hints.SUPPORTABLE,
+                            ],
+                        )
+                elif isinstance(var, BaseListVariable):
+                    for item in var.items:
+                        check_lazy_tensor_grad_fn(item)
+
+            for arg in args[:2]:  # outputs and inputs
+                check_lazy_tensor_grad_fn(arg)
+
+            for source, var in tx.output.input_source_to_var.items():
+                if source in tx.output.autograd_grad_checked_sources:
+                    continue
+                tx.output.autograd_grad_checked_sources.add(source)
+
+                if isinstance(var, TensorVariable) and var.has_grad_fn:
+                    unimplemented(
+                        gb_type="autograd.grad with external grad_fn (input)",
+                        context="",
+                        explanation=(
+                            "torch.autograd.grad() cannot be used when the compiled function "
+                            "receives tensors with grad_fn created outside the compiled region. "
+                            "The autograd graph cannot extend beyond the compiled region boundary. "
+                            "Note: This is a conservative check. It may be safe if the input with "
+                            "grad_fn is not reachable from the autograd graph of the differentiated tensors."
```

Comment:
Honestly, let's just do this properly with a `grad_fn` traversal, I think...

Reply:
Yep done.